src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java

```diff
@@ -0,0 +1,225 @@
+
+    public AbstractSecurityAnalysis(Network network, LimitViolationDetector detector, LimitViolationFilter filter,
+                                    MatrixFactory matrixFactory, Supplier<GraphDecrementalConnectivity<LfBus>> connectivityProvider) {
+        this.network = Objects.requireNonNull(network);
+        this.detector = Objects.requireNonNull(detector);
+        this.filter = Objects.requireNonNull(filter);
+        this.matrixFactory = Objects.requireNonNull(matrixFactory);
+        this.connectivityProvider = Objects.requireNonNull(connectivityProvider);
+    }
+
+    @Override
+    public void addInterceptor(SecurityAnalysisInterceptor interceptor) {
+        interceptors.add(Objects.requireNonNull(interceptor));
+    }
+
+    @Override
+    public boolean removeInterceptor(SecurityAnalysisInterceptor interceptor) {
+        return interceptors.remove(Objects.requireNonNull(interceptor));
+    }
+
+    @Override
+    public CompletableFuture<SecurityAnalysisResult> run(String workingVariantId, SecurityAnalysisParameters securityAnalysisParameters, ContingenciesProvider contingenciesProvider) {
+        Objects.requireNonNull(workingVariantId);
+        Objects.requireNonNull(securityAnalysisParameters);
+        Objects.requireNonNull(contingenciesProvider);
+        return CompletableFuture.supplyAsync(() -> {
+            String oldWorkingVariantId = network.getVariantManager().getWorkingVariantId();
+            network.getVariantManager().setWorkingVariant(workingVariantId);
+            SecurityAnalysisResult result = runSync(securityAnalysisParameters, contingenciesProvider);
+            network.getVariantManager().setWorkingVariant(oldWorkingVariantId);
+            return result;
+        });
+    }
+
+    abstract SecurityAnalysisResult runSync(SecurityAnalysisParameters securityAnalysisParameters, ContingenciesProvider contingenciesProvider);
+
+    /**
+     * Detect violations on branches and on buses
+     * @param branches branches on which the violation limits are checked
+     * @param buses buses on which the violation limits are checked
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectViolations(Stream<LfBranch> branches, Stream<LfBus> buses, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // Detect violation limits on branches
+        branches.forEach(branch -> detectBranchViolations(branch, violations));
+
+        // Detect violation limits on buses
+        buses.forEach(bus -> detectBusViolations(bus, violations));
+    }
+
+    /**
+     * Detect violation limits on one branch and add them to the given list
+     * @param branch branch of interest
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectBranchViolations(LfBranch branch, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // detect violation limits on a branch
```

Comment:
Maybe add here the comment that we always take only the most serious violation, and as the limits are sorted in descending gravity that's why we take the first one